File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        unordered_map<int,int> mp; // value - index;
        // find greatest element
        stack<pair<int,int>> s;
        for(int i = 0; i<nums2.size(); i++){
            mp.insert({nums2[i],i});
            int curr = nums2[i];
            while(!s.empty() && curr>s.top().first){
                nums2[s.top().second] = curr;
                s.pop();
            }
            s.push({curr,i});  
        }
        while(!s.empty()){
            nums2[s.top().second] = -1;
            s.pop();
        }
        for(int i = 0; i<nums1.size(); i++){
            nums1[i] = nums2[mp[nums1[i]]];
        }
        return nums1;
    }
};

File: 0496-next-greater-element-i/0496-next-greater-element-i.js
// /**
//  * @param {number[]} nums1
//  * @param {number[]} nums2
//  * @return {number[]}
//  */
// var nextGreaterElement = function(nums1, arr) {
//     const n = arr.length;
//     let ans = new Array(n).fill(-1);
//         let stack = []; // form [ [value,index] ]
//         for(let i = 0; i<n; i++){
//             const curr = arr[i];
//             while(stack.length && curr>stack.at(-1)[0]){
//                 ans[stack[stack.length-1][1]] = curr;
//                 stack.pop();
//             }
//             stack.push([curr,i]);
//         }
//     return ans;
// };
/**
 * @param {number[]} nums1
 * @param {number[]} nums2
 * @return {number[]}
 */
var nextGreaterElement = function(nums1, arr) {
    const n = arr.length;
    const temp = new Array(n).fill(-1); // for grater element 
    const map = new Map();
    const stack = []; // [ [value] index ]
    for(let i = 0; i<n; i++){
        const curr = arr[i];
        map.set(curr,i)
        while(stack.length && curr > stack.at(-1)[0]) temp[stack.pop()[1]] = curr;
        stack.push([curr,i]);
    }
    for(let i = 0; i<nums1.length; i++) nums1[i] = temp[map.get(nums1[i])];
    return nums1;
};